samples/cellular/nrf_cloud_multi_service/src/provisioning_support.c

```diff
@@ -0,0 +1,121 @@
+
+#define NETWORK_UP			BIT(0)
+#define NETWORK_DOWN			BIT(1)
+
+static K_EVENT_DEFINE(prov_events);
+static bool provisioning_started = false;
+
+/* Called by the provisioning library to request the LTE modem be taken offline before credential
+ * installation, and then again after credential installation is complete. This is necessary because
+ * credentials cannot be installed to the modem while it is online.
+ *
+ * This limitation only applies to the nrf91 modem; Wi-Fi, for instance, is unaffected. But for
+ * simplicity we use conn_mgr to take take all network interfaces offline and online.
+ */
+static int modem_mode_cb(enum lte_lc_func_mode new_mode, void *user_data)
+{
+	enum lte_lc_func_mode fmode;
+
+	ARG_UNUSED(user_data);
+
+	/* The nrf_provisioning library requires us to return the previous functional mode. */
+	if (lte_lc_func_mode_get(&fmode)) {
+		LOG_ERR("Failed to read modem functional mode");
+		return -EFAULT;
+	}
+
+	if (new_mode == LTE_LC_FUNC_MODE_NORMAL || new_mode == LTE_LC_FUNC_MODE_ACTIVATE_LTE) {
+		LOG_DBG("Provisioning library requests normal mode");
+
+		/* Connect all interfaces since we are done provisioning. */
+		conn_mgr_all_if_connect(true);
+
+		/* Wait for network readiness to be re-established before returning. */
+		LOG_DBG("Waiting for network up");
+		k_event_wait(&prov_events, NETWORK_UP, false, K_FOREVER);
+
+		LOG_DBG("Network is up.");
+	} else {
+		LOG_DBG("Provisioning library requests offline mode");
+		/* Disconnect all interfaces to allow cert installation. */
+		conn_mgr_all_if_disconnect(true);
+
+		/* Wait for network disconnection before returning. */
+		LOG_DBG("Waiting for network down");
+		k_event_wait(&prov_events, NETWORK_DOWN, false, K_FOREVER);
+
+		LOG_DBG("Network is down.");
+	}
+
+	return fmode;
+}
+static struct nrf_provisioning_mm_change mmode = { .cb = modem_mode_cb };
+
+/* Initialize the provisioning library and start checking for provisioning commands.
+ * Called automatically the first time network connectivity is established.
+ */
+void start_provisioning(void)
+{
+	await_date_time_known(K_FOREVER);
+	LOG_INF("Initializing the nRF Provisioning library...");
+
+	/*
+	 * Pass NULL for dmode callback, since the default one is perfectly fine.
+	 * The default callback will automatically reboot after provisioning completes.
+	 */
+	int ret = nrf_provisioning_init(&mmode, NULL);
+
+	if (ret) {
+		LOG_ERR("Failed to initialize provisioning client, error: %d", ret);
+	}
+}
+
+/* Callback to track network connectivity */
+static struct net_mgmt_event_callback l4_callback;
+static void l4_event_handler(struct net_mgmt_event_callback *cb,
+			     uint32_t event, struct net_if *iface)
+{
+	if (event == NET_EVENT_L4_CONNECTED) {
+		/* Start the provisioning library after network readiness is first established. */
+		if (!provisioning_started) {
+			start_provisioning();
+			provisioning_started = true;
+		}
+
+		k_event_post(&prov_events, NETWORK_UP);
```

Comment:
should the events be updated before provisioning is started?

Reply:
It doesn't matter too terribly much since the provisioning client spools up its own thread, but I may as well update the events first, sure